### PR TITLE
Cesarvr/mixed for linked realm object

### PR DESCRIPTION
**Pull request by cesarvr**
This PR adds [Typed_Link](https://realm.io/docs/javascript/latest/#relationships) support to Realm-JS. 
Allowing the user to use the Mixed field type to store this relationships and primitive types.

### CODE REVIEW

File: src/common/mixed_type.hpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
```
Comment (kneth) [inline src/common/mixed_type.hpp:3]: `2016` -> `2021`

File: src/js_links.hpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
```
Comment (kneth) [inline src/js_links.hpp:3]: `2016`-> `2021`

File: src/js_links.hpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+template <typename T>
+class MixedLink : public MixedWrapper<typename T::Context, typename T::Value> {
+   private:
+    using Context = typename T::Context;
+    using Value = typename T::Value;
+    using RealmClass =  RealmObjectClass<T>;
+
+    std::shared_ptr<Realm> realm;
```
Comment (kneth) [inline src/js_links.hpp:71]: We often have `using SharedRealm = std::shared_ptr<Realm>` - saves many keystrokes :-)

File: tests/js/mixed-tests.js [hunk 4]
```diff
@@ -83,12 +85,75 @@ module.exports = {
...
+   
+        realm.write(() => {
+           let r =  realm.create(VertexSchema.name,{a:1, b: 0,  c: 0})
+           let r2 = realm.create(VertexSchema.name,{a:0, b: 1,  c: 0})
+           let r3 = realm.create(VertexSchema.name,{a:0, b: 0,  c: 1})
+
+           realm.create(SingleSchema.name, { a: r, b: r2, c: r3 } ) 
+        })
+
+        let data = realm.objects(SingleSchema.name)[0]
```
Comment (kneth) [inline tests/js/mixed-tests.js:119]: You might want to assert the length is 1 before accessing the element.

File: tests/js/mixed-tests.js [hunk 5]
```diff
@@ -83,12 +85,75 @@ module.exports = {
...
+        TestCase.assertEqual(data.c.b, 0, 'Should be equal 0');
+        TestCase.assertEqual(data.c.c, 1, 'Should be equal 1');
+
+        realm.write(() => {
+           let r =  realm.create(MixNestedSchema.name,{a:0, b: -1})
+           let r1 = realm.create(MixNestedSchema.name,{a:1, b: 0})
+           realm.create(SingleSchema.name, { a: r, b: r1 } ) 
+        })
+
+        data = realm.objects(SingleSchema.name)[1]
```
Comment (kneth) [inline tests/js/mixed-tests.js:138]: You might want to assert the length is 2 before accessing the element.

File: src/js_links.hpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+        RealmLink<T> realm_link {context, value};
+
+        if(!realm_link.is_instance() ||
+        !realm_link.belongs_to_realm(realm)){
+            throw std::runtime_error("Only Realm objects are supported.");
+        }
+
+        auto realm_object = realm_link.get_realm_object();
+        return Mixed(realm_object);
```
Comment (nirinchev) [inline src/js_links.hpp:86]: `std::move` this or combine both lines in a single one.

File: src/js_object_accessor.hpp [hunk 7]
```diff
@@ -421,33 +434,33 @@ struct Unbox<JSEngine, Timestamp> {
...
-            if (!updating && !policy.create) {
-                throw std::runtime_error("Realm object is from another Realm");
-            }
+        if(realm_link.belongs_to_realm(current_realm)){
+            return realm_link.get_realm_object();
+        }
+
+        if(realm_link.is_instance() && realm_link.is_read_only(policy)){
+            throw std::runtime_error("Realm object is from another Realm");
         }
```
Comment (nirinchev) [inline src/js_object_accessor.hpp:450]: Why does the fact that a realm link is read only imply that it belongs to another Realm? Either the naming of the method is off or the error message is.

File: src/js_object_accessor.hpp [hunk 8]
```diff
@@ -421,33 +434,33 @@ struct Unbox<JSEngine, Timestamp> {
...
+        auto object = Value::validated_to_object(native_accessor->m_ctx, value);
+        if (Value::is_array(native_accessor->m_ctx, object)) {
+            object = Schema<JSEngine>::dict_for_property_array(native_accessor->m_ctx, *native_accessor->m_object_schema, object);
         }
 
-        auto child = realm::Object::create<ValueType>(*ctx, ctx->m_realm, *ctx->m_object_schema,
-                                                      static_cast<ValueType>(object), policy, current_row);
+        auto child = realm::Object::create<ValueType>
+            (*native_accessor, native_accessor->m_realm, *native_accessor->m_object_schema,
+                    static_cast<ValueType>(object), policy, current_row);
```
Comment (nirinchev) [inline src/js_object_accessor.hpp:463]: This formatting looks off/inconsistent with other files - does realm-js not have clang or something similar to enforce consistent code style?

**Review by nirinchev — COMMENTED** [review]
I have mostly code style + error message comments. Also, seems to be missing tests for trying to assign a property to an object instance that belongs to a different Realm and trying to delete an object and then assign it to a mixed property.

File: src/js_object_accessor.hpp (hunk 8 above)
Comment (kraenhansen) [reply src/js_object_accessor.hpp:463]: We don't have that yet and until we have, I suggest not focussing too much on code style when reviewing PRs to the Realm JS C++ code. We will the formatter on the entire codebase once we get there.

**Review by fronck — APPROVED** [review]
LGTM

**Review by fronck — APPROVED** [review]

**Pull request merged**